File: Red_Black_Trees/rbt_rotate.py
# File: rbt_rotate.py

class Node:
    def __init__(self, data):
        self.data = data
        self.color = "RED"
        self.left = None
        self.right = None
        self.parent = None


class RedBlackTree:

    def __init__(self):
        self.root = None

    def left_rotate(self, x):
        y = x.right

        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def right_rotate(self, y):
        x = y.left

        y.left = x.right
        if x.right:
            x.right.parent = y

        x.parent = y.parent

        if not y.parent:
            self.root = x
        elif y == y.parent.left:
            y.parent.left = x
        else:
            y.parent.right = x

        x.right = y
        y.parent = x


if __name__ == "__main__":
    print("Red-Black Tree rotation operations implemented successfully.")

File: Red_Black_Trees/rbt_rotate.c
/* File: rbt_rotate.c */

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    int color;   // 0 = RED, 1 = BLACK
    struct Node *left, *right, *parent;
} Node;

Node* createNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->color = 0;
    node->left = node->right = node->parent = NULL;
    return node;
}

void leftRotate(Node **root, Node *x) {
    Node *y = x->right;

    x->right = y->left;
    if (y->left != NULL)
        y->left->parent = x;

    y->parent = x->parent;

    if (x->parent == NULL)
        *root = y;
    else if (x == x->parent->left)
        x->parent->left = y;
    else
        x->parent->right = y;

    y->left = x;
    x->parent = y;
}

void rightRotate(Node **root, Node *y) {
    Node *x = y->left;

    y->left = x->right;
    if (x->right != NULL)
        x->right->parent = y;

    x->parent = y->parent;

    if (y->parent == NULL)
        *root = x;
    else if (y == y->parent->left)
        y->parent->left = x;
    else
        y->parent->right = x;

    x->right = y;
    y->parent = x;
}

int main() {

    Node *root = createNode(20);
    root->left = createNode(10);
    root->left->parent = root;

    printf("Rotation functions implemented successfully.\n");

    return 0;
}
